[master] Allow opt-out of discard on disk format

jseba opened the pull request:
### What does this PR do?
This allows disk images to be formatted explicitly non-sparsely by enabling callers to disable the automatic discard of blocks during filesystem creation.

This is only supported for ext and xfs filesystems.

### What issues does this PR fix or reference?
Fixes:

Current behavior of formatting disk images results in e.g. a 20GB disk image being reduced to 44MB on the parent filesystem, which shows up as potential free space when other things are looking at the parent's disk usage. If the parent filesystem no longer has blocks to service the disk image's usage, it gets ENOSPC errors when it is not at 100% usage.

### New Behavior
Callers to `disk.format` and `blockdev.formatted` can specify `discard=False` for ext and xfs filesystems to disable their respective mkfs tools from doing a block discard during filesystem creation. This ensures the disk image is still reserving blocks on the parent filesystem for itself if it was created non-sparsely.

### Merge requirements satisfied?
- [X] Docs
- [ ] Changelog
- [X] Tests written/updated

### Commits signed with GPG?
No

Ch3LL reviewed (changes requested):
This looks good, but just needs a changelog entry (https://docs.saltproject.io/en/latest/topics/development/changelog.html)

jseba commented:
I added the changelog entry, also went ahead and rebased on master

Ch3LL commented:
Looks like there are related test failures that need to be resolved. Also can you ensure to rebase as there was recent test fixes merged in.

twangboy commented on salt/modules/disk.py:
```diff
@@ -475,6 +476,11 @@ def format_(
 
         This option is dangerous, use it with caution.
 
+    discard
```
This needs a `.. version-added::` marker

twangboy reviewed (changes requested):
This also needs to be rebased and conflicts resolved

dwoz committed:
Allow opt-out of discard on disk format

Add a ``discard`` parameter (default ``True``) to ``disk.format_`` so
callers can disable block discard during filesystem creation.  Without
this, mkfs discards all blocks in a disk image, collapsing a 20 GB
pre-allocated image to ~44 MB on the host filesystem and causing ENOSPC
errors when the host runs low on space.

For ext filesystems the flag maps to ``-E nodiscard``; for xfs it maps
to ``-K``.  The ``blockdev.formatted`` state already forwards ``**kwargs``
to ``disk.format_``, so passing ``discard=False`` in state works without
any state-module change.

Fixes #64310
